src/null-safety/understanding-null-safety/index.md

```diff
@@ -0,0 +1,1410 @@
+
+```dart
+bool isEmptyList(Object object) {
+  if (object is! List) return false;
+  return object.isEmpty; // <-- Error!
+}
+```
+
+Again, you can only reach the `.isEmpty` call when `object` contains a list, so
+this program is dynamically correct. But the type promotion rules are not smart
+enough to see that the `return` statement means the second statement can only be
+reached when `object` is a list.
+
+For null safety, we've taken this limited analysis and made it [much more
+powerful in several ways][flow analysis].
+
+[flow analysis]: https://github.com/dart-lang/language/blob/master/resources/type-system/flow-analysis.md
+
+### Reachability analysis
+
+First off, we fixed the [long-standing complaint][18921] that type promotion
+isn't smart about early returns an other unreachable code paths. When analyzing
+a function, it now takes into account `return`, `break`, `throw`, and any other
+way execution might terminate early in a function. Under null safety, this function:
+
+[18921]: https://github.com/dart-lang/sdk/issues/18921
+
+```dart
+bool isEmptyList(Object object) {
+  if (object is! List) return false;
+  return object.isEmpty;
+}
+```
+
+Is now perfectly valid. Since the `if` statement will exit the function when
+`object` is *not* a `List`, Dart promotes `object` to be `List` on the second
+statement. This is a really nice improvement that helps a lot of Dart code, even
+stuff not related to nullability.
+
+### Never for unreachable code
+
+You can also *program* this reachability analysis. The new bottom type `Never`
+has no values. (What kind of value is simultaneously a `String`, `bool`, and
+`int`?) So what does it mean for an expression to have type `Never`? It means
+that expression can never successfully finish evaluating. It must throw an
+exception, abort, or otherwise ensure that the surrounding code expecting the
+result of the expression never runs.
+
+In fact, according to the language, the static type of a `throw` expression is
+`Never`. The type `Never` is declared in the core libraries and you can use it
+as a type annotation. Maybe you have a helper function to make it easier to
+throw a certain kind of exception:
+
+```dart
+Never wrongType(String type, Object value) {
+  throw ArgumentError('Expected $type, but was ${value.runtimeType}.');
+}
+```
+
+You might use it like so:
+
+```dart
+class Point {
+  final double x, y;
+
+  bool operator ==(Object other) {
+    if (other is! Point) wrongType("Point", other);
+    return x == other.x && y == other.y;
+  }
+
+  // Constructor and hashCode...
+}
+```
+
+This program analyzes without error. Notice that the last line of the `==`
+method accesses `.x` and `.y` on `other`. It has been promoted to `Point` even
+though the function doesn't have any `return` or `throw`. The control flow
+analysis knows that the declared type of `wrongType()` is `Never` which means
+the then body of the `if` statement *must* abort somehow. Since the second
```

Comment:
OOOOOHH, now i get it. Yes, please code font.  :)

Reply:
Changed to "then branch" like the other places use. Putting it in code font seems weird to me since "then" isn't a keyword in Dart. I could hyphenate like "then-branch" if you think that helps. A couple of textbooks I have do that to refer to programming language constructs.